synapse/tests/test_lib_remcycle.py

```diff
@@ -0,0 +1,623 @@
+        hypo_obj = s_remcycle.Hypnos(ioloop=self.io_loop)
+        hypo_obj.register_config(config=gconf)
+        core = s_cortex.openurl('ram://')
+
+        data = {}
+
+        ingest_def = {
+            "ingest": {
+                "forms": [
+                    [
+                        "inet:ipv4",
+                        {
+                            "var": "ip"
+                        }
+                    ]
+                ],
+                "vars": [
+                    [
+                        "ip",
+                        {
+                            "path": "ip"
+                        }
+                    ]
+                ]
+            }
+        }
+
+        name = 'ipify:jsonip'
+        core_name = ':'.join([name, 'ingest'])
+
+        gest = s_ingest.Ingest(info=ingest_def)
+
+        s_ingest.register_ingest(core=core,
+                                 gest=gest,
+                                 evtname=core_name)
+
+        def glue(event):
+            evtname, event_args = event
+            kwargs = event_args.get('kwargs')
+            resp = kwargs.get('resp')
+            data = resp.get('data')
+            core.fire(core_name, data=data)
+
+        def ondone(job_tufo):
+            _jid, jobd = job_tufo
+            ip = jobd.get('task')[2].get('resp', {}).get('data', {}).get('ip', '')
+            data[_jid] = ip
+
+        hypo_obj.on(name=name, func=glue)
+
+        jid = hypo_obj.fire_api(name=name, ondone=ondone)
+        hypo_obj.boss.wait(jid)
+
+        tufos = core.getTufosByProp('inet:ipv4')
+        self.eq(len(tufos), 1)
+        # Validate the IP of the tufo is the same we got from ipify
+        self.eq(s_inet.ipv4str(tufos[0][1].get('inet:ipv4')), data[jid])
+
+        hypo_obj.fini()
+        core.fini()
+
+    def test_hypnos_automatic_ingest(self):
+        """
+        Ensure that a configuration object with a ingest definition is automatically parsed.
+        """
+        self.skipIfNoInternet()
+        gconf = get_ipify_ingest_global_config()
+        hypo_obj = s_remcycle.Hypnos(ioloop=self.io_loop)
+        hypo_obj.register_config(config=gconf)
+
+        self.true('ipify:jsonip' in hypo_obj._syn_funcs)
+        self.true('ipify:jsonip:ipv4' in hypo_obj.core._syn_funcs)
+
+        data = {}
+
+        def ondone(job_tufo):
+            _jid, jobd = job_tufo
+            ip = jobd.get('task')[2].get('resp', {}).get('data', {}).get('ip', '')
+            data[_jid] = ip
+
+        jid = hypo_obj.fire_api(name='ipify:jsonip', ondone=ondone)
+        hypo_obj.boss.wait(jid)
+
+        tufos = hypo_obj.core.getTufosByProp('inet:ipv4')
+        self.eq(len(tufos), 1)
+        # Validate the IP of the tufo is the same we got from ipify
+        self.eq(s_inet.ipv4str(tufos[0][1].get('inet:ipv4')), data[jid])
+
+        hypo_obj.fini()
+
+    def test_hypnos_throw_timeouts(self):
+        """
+        Run a test scenario which will generate hundreds of jobs which will timeout.
```

Comment:
This test doesn't consistently pass from run to run.  I've marked it for skipping.